### PR TITLE
mgr/rgw: Adding a retry config while calling zone_create()

### PR DESCRIPTION
**Pull request by skritik098**
Fixes https://tracker.ceph.com/issues/66750

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by gauravsitlani** [comment]
Jenkins retest this please

File: src/pybind/mgr/rgw/module.py [hunk 1]
```diff
@@ -101,7 +101,14 @@ def wrapper(self: Any, *args: Any, **kwargs: Any) -> HandleCommandResult:
 
 
 class Module(orchestrator.OrchestratorClientMixin, MgrModule):
-    MODULE_OPTIONS: List[Option] = []
+    MODULE_OPTIONS: List[Option] = [
+         Option(
```
Comment (adk3798) [inline src/pybind/mgr/rgw/module.py:105]:
```
rgw/module.py:105:10: E126 continuation line over-indented for hanging indent
```

File: src/pybind/mgr/rgw/module.py [hunk 2]
```diff
@@ -101,7 +101,14 @@ def wrapper(self: Any, *args: Any, **kwargs: Any) -> HandleCommandResult:
...
+            'rgw_retry_limit',
```
Comment (adk3798) [inline src/pybind/mgr/rgw/module.py:106]:
```
rgw/module.py:106:13: E121 continuation line under-indented for hanging indent
```

File: src/pybind/mgr/rgw/module.py [hunk 3]
```diff
@@ -101,7 +101,14 @@ def wrapper(self: Any, *args: Any, **kwargs: Any) -> HandleCommandResult:
...
+            type='int',
+            default=5,
+            desc='RGW module period update retry limit for secondary site',
+        ),
```
Comment (adk3798) [inline src/pybind/mgr/rgw/module.py:110]:
```
rgw/module.py:110:9: E122 continuation line missing indentation or outdented
```

File: src/pybind/mgr/rgw/module.py [hunk 4]
```diff
@@ -312,7 +322,7 @@ def _cmd_rgw_zone_create(self,
 
         try:
             created_zones = self.rgw_zone_create(zone_name, realm_token, port, placement,
-                                                 start_radosgw, zone_endpoints, inbuf)
+                                                 start_radosgw, zone_endpoints,self.rgw_retry_limit, inbuf)
```
Comment (adk3798) [inline src/pybind/mgr/rgw/module.py:325]:
```
rgw/module.py:325:79: E231 missing whitespace after ','
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 5]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
         zone = self.create_zone(realm, zonegroup, rgw_spec.rgw_zone,
                                 False,  # secondary zone
                                 access_key, secret, endpoints=rgw_spec.zone_endpoints)
-        self.update_period(realm, zonegroup, zone)
+
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:826]:
```
ceph/rgw/rgwam_core.py:826:1: W293 blank line contains whitespace
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 6]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
...
                                 False,  # secondary zone
                                 access_key, secret, endpoints=rgw_spec.zone_endpoints)
-        self.update_period(realm, zonegroup, zone)
+        
+        # Adding a retry limit for period update in case the default 10s timeout is not sufficient
+
+        # Modify this section of the code for Delay in period update ##
+        rgw_limit = int(0)
+
+        while rgw_limit!=int(rgw_retry_limit):
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:832]:
```
ceph/rgw/rgwam_core.py:832:24: E225 missing whitespace around operator
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 7]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
...
+            try:
+                self.update_period(realm, zonegroup, zone)
+                break
+            except:
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:836]:
```
ceph/rgw/rgwam_core.py:836:13: E722 do not use bare 'except'
```

Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:836]:
```
ceph/rgw/rgwam_core.py:836:20: W291 trailing whitespace
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 8]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
...
+
+        # Modify this section of the code for Delay in period update ##
+        rgw_limit = int(0)
+
+        while rgw_limit!=int(rgw_retry_limit):
+            try:
+                self.update_period(realm, zonegroup, zone)
+                break
+            except: 
+                logging.info(f'Failed to update Period in 10s timeframe. Retrying it again while the current limit count {rgw_limit} & retry-limit value of {rgw_retry_limit}')
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:837]:
```
ceph/rgw/rgwam_core.py:837:101: E501 line too long (175 > 100 characters)
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 9]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
...
+                rgw_limit = rgw_limit+1
+                if rgw_limit == rgw_retry_limit:
+                    raise RGWAMException(f'Period Update failed for zonegroup {zonegroup} & zone {zone}')
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:840]:
```
ceph/rgw/rgwam_core.py:840:101: E501 line too long (105 > 100 characters)
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 10]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
...
+                continue
+
+        # By default the above operation is expected to be completed in 10s timeout
+        # but if we updating this for secondary site it would take some time because of pool creation
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:844]:
```
ceph/rgw/rgwam_core.py:844:101: E501 line too long (101 > 100 characters)
```

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 11]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
         zone = self.create_zone(realm, zonegroup, rgw_spec.rgw_zone,
                                 False,  # secondary zone
                                 access_key, secret, endpoints=rgw_spec.zone_endpoints)
-        self.update_period(realm, zonegroup, zone)
+        
+        # Adding a retry limit for period update in case the default 10s timeout is not sufficient
+
+        # Modify this section of the code for Delay in period update ##
+        rgw_limit = int(0)
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:830]: I don't think we need the `int()` around the 0

File: src/python-common/ceph/rgw/rgwam_core.py [hunk 12]
```diff
@@ -823,7 +823,25 @@ def zone_create(self, rgw_spec, start_radosgw):
...
+
+        while rgw_limit!=int(rgw_retry_limit):
+            try:
+                self.update_period(realm, zonegroup, zone)
+                break
+            except: 
+                logging.info(f'Failed to update Period in 10s timeframe. Retrying it again while the current limit count {rgw_limit} & retry-limit value of {rgw_retry_limit}')
+                rgw_limit = rgw_limit+1
```
Comment (adk3798) [inline src/python-common/ceph/rgw/rgwam_core.py:838]: nitpick: this could be `rgw_limit += 1`

**Review by adk3798 — COMMENTED** [review]
mostly complaints from the linter that have to be fixed for the required checks to pass

File: src/pybind/mgr/rgw/module.py (hunk 2 above)
Comment (adk3798) [inline src/pybind/mgr/rgw/module.py:106]: I feel like we ought to have a more specific name for this setting given it's only being used for period updates during zone creation

Comment (skritik098) [reply src/pybind/mgr/rgw/module.py]: @adk3798 I have changed the parameter name to "secondary_zone_period_retry_limit", does this look good ?

Please let me know if you have any suggestions or feedback.

**Comment by adk3798** [comment]
@skritik098 I think a mistake was made while updating the PR. There are now a bunch of changes to crimson code included.

**Comment by skritik098** [comment]
@adk3798 I have reverted all of the changes that comes from black utility for having double quotes instead of single one.

Please let me know your feedback on the parameter name and any other place a change is required.

**Comment by adk3798** [comment]
jenkins test make check

**Comment by gauravsitlani** [comment]
jenkins test make check

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2024-07-09_19:40:48-orch:cephadm-wip-adk-testing-2024-07-09-0905-distro-default-smithi/

Failures:

- mds_upgrade_sequence, known issue
- staggered upgrade with agent, known issue
- a couple `in cluster logs` type failures that don't look bad
- one failure in staggered upgrade that seemed to be a mismatch between what `ceph versions` and cephadm believed to be the version for a single RGW daemon. Needs investigation, but unrelated to PRs in the run
- thrash and rgw-ingress tests timing out, known issue

Generally a good run, nothing to block merging most of the PRs

**Pull request merged**